Source/WebCore/html/ColorInputType.cpp

```diff
@@ -86,6 +93,56 @@ static std::optional<SRGBA<uint8_t>> parseSimpleColorValue(StringView string)
     return { { toASCIIHexValue(string[1], string[2]), toASCIIHexValue(string[3], string[4]), toASCIIHexValue(string[5], string[6]) } };
 }
 
+static LazySlowPathColorParsingParameters colorParsingParameters()
+{
+    return {
+        CSSPropertyParserHelpers::CSSColorParsingOptions {
+            .allowedColorTypes = { StyleColor::CSSColorType::Absolute, StyleColor::CSSColorType::Current, StyleColor::CSSColorType::System }
+        },
+        CSSUnresolvedColorResolutionState {
+            .resolvedCurrentColor = Color::black
+        },
+        std::nullopt
+    };
+}
+
+static std::optional<Color> parseColorValue(StringView string, HTMLInputElement& context)
+{
+    if (context.colorSpace().isNull())
+        return parseSimpleColorValue(string);
+
+    auto color = CSSPropertyParserHelpers::parseColorRaw(string.toString(), CSSParserContext(HTMLStandardMode), [&] {
+        return colorParsingParameters();
+    });
+
+    if (!color.isValid())
+        return { };
+
+    return color;
+}
+
+static String serializeColorValue(Color input, HTMLInputElement& context)
+{
+    auto colorSpace = context.colorSpace();
+    auto alpha = !colorSpace.isNull() && context.alpha();
+
+    if (!alpha)
+        input = input.opaqueColor();
+
+    if (colorSpace.isNull() || colorSpace == "limited-srgb"_s) {
+        auto inputAsRGBA = input.toColorTypeLossy<SRGBA<uint8_t>>();
+        if (alpha)
+            input = { inputAsRGBA, { Color::Flags::UseColorFunctionSerialization } };
+        else
+            input = inputAsRGBA.resolved();
+    } else {
+        // FIXME: Ideally this would not be a lossy conversion and preserve colors such as color(display-p3 2 2 2).
+        input = input.toColorTypeLossy<DisplayP3<float>>().resolved();
```

Comment:
@weinig I'd appreciate your input on this FIXME and more generally on the integration with the color code if you have the time. Thanks!

Reply:
The name is a bit misleading. The conversion is only lossy if the requested color type would cause it to be lossy. So it's more like, `toColorType*Potentially*Lossy()`.

In this case though, you should use `ExtendedDisplayP3` rather than `DisplayP3`, so there is no clamping applied.

--

I would probably put this in an else if (colorSpace == "display-p3") block and/or assert that is the color space, as its a bit tricky to locally reason about why using DisplayP3 here is right.